libraries/standard/http/include/http_client.h

```diff
@@ -0,0 +1,339 @@
+ * header. 
+ */
+#define HTTP_REQUEST_DISABLE_AUTO_CONTENT_LENGTH_FLAG   0x2U
+
+
+/**
+ * Flags for #HTTPResponse_t.flags.
+ * These flags are in #HTTPResponse_t.flags after #HTTPClient_Send() returns.
+ */
+ /**
+  * @brief This will be set to true if header "Connection: close" is found.
+  * 
+  * If a "Connection: close" header is present the client should always
+  * close the connection.
+  */
+#define HTTP_RESPONSE_CONNECTION_CLOSE_FLAG             0x1U
+/**
+ * @brief This will be set to true if header "Connection: Keep-Alive" is found.
+ */
+#define HTTP_RESPONSE_CONNECTION_KEEP_ALIVE_FLAG        0x2U
+
+/**
+ * @brief The HTTPNetworkContext is an incomplete type. The application must 
+ * define HTTPNetworkContext to the type of their network context. This context
+ * is passed into the network interface functions.
+ */
+struct HTTPNetworkContext;
+typedef struct HTTPNetworkContext * HTTPNetworkContext_t;
+
+/**
+ * @brief Transport interface for sending data over the network.
+ * 
+ * If the number of bytes written returned is less than bytesToWrite, then 
+ * #HTTPClient_Send will return HTTP_NETWORK_ERROR. If a negative value is 
+ * returned then this #HTTPClient_Send will also return #HTTP_NETWORK_ERROR.
+ * 
+ * @param[in] context User defined context.
+ * @param[in] pBuffer Buffer to write to the network stack.
+ * @param[in] bytesToWrite Number of bytes to write to the network.
+ * @return The number of bytes written or a negative network error code.
+ */
+typedef int32_t (* HTTPTransportSend_t )( HTTPNetworkContext_t context,
+                                          const void * pBuffer, 
+                                          size_t bytesToWrite );
+
+/**
+ * @brief Transport interface for reading data on the network.
+ * 
+ * This function will read up to bytesToRead amount of data from the network.
+ * If this function returns a value less than zero, then #HTTPClient_Send will
+ * return #HTTP_NETWORK_ERROR. If this function returns less than the 
+ * bytesToRead and greater than zero, then this function will be invoked again
+ * if the data in pBuffer contains a partial HTTP response message.
+ * 
+ * @param[in] context User defined context.
+ * @param[in] pBuffer Buffer to read network data into.
+ * @param[in] bytesToRead Number of bytes requested from the network.
+ * @return The number of bytes read or a negative error code.
+ */
+typedef int32_t (* HTTPTransportRecv_t )( HTTPNetworkContext_t context,
+                                          const void * pBuffer, 
+                                          size_t bytesToRead );
+
+/**
+ * @brief The HTTP Client library transport layer interface.
+ */
+typedef struct HTTPTransportInterface
+{
+    HTTPTransportRecv_t recv;
+    HTTPTransportSend_t send;
+    HTTPNetworkContext_t context;
+} HTTPTransportInterface_t;
+
+/**
+ * @brief The HTTP Client library return status.
+ */
+typedef enum HTTPStatus
+{
+    HTTP_SUCCESS = 0,
```

Comment:
@dan4thewin may agree with me generally on not defining `SUCCESS` status as 0 to avoid accidental happy-path execution when dealing with _zero-initialized static_ data

Reply:
Let's discuss with the MQTT developers as well.

Reply:
Hard to say when 0 generally means no error when 200 is returned by website upon success so it's a good discussion to be mode.

Reply:
I am ok with changing this, but that can be done later for both libraries. These statuses are not linked the HTTP response status code.